compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs

```diff
@@ -63,6 +92,353 @@ fn check_for_underconstrained_values_within_function(
+                        | Value::NumericConstant { .. }
+                        | Value::Param { .. } => {
+                            panic!(
+                                "should not be able to call {func_id} in function {}",
+                                function.name()
+                            );
+                        }
+                    }
+                }
+                // For array get operations, we link the resulting values to
+                // the corresponding array value ids
+                // (this is required later because for now we consider array elements
+                // being constrained as valid as the whole arrays being constrained)
+                Instruction::ArrayGet { array, .. } => {
+                    for result in &results {
+                        self.array_elements.insert(*result, function.dfg.resolve(*array));
+                    }
+                    // Record all the used arguments as parents of the results
+                    self.update_children(&arguments, &results);
+                }
+                Instruction::ArraySet { .. }
+                | Instruction::Binary(..)
+                | Instruction::Cast(..)
+                | Instruction::IfElse { .. }
+                | Instruction::Not(..)
+                | Instruction::Truncate { .. } => {
+                    // Record all the used arguments as parents of the results
+                    self.update_children(&arguments, &results);
+                }
+                // These instructions won't affect the dependency graph
+                Instruction::Allocate { .. }
+                | Instruction::DecrementRc { .. }
+                | Instruction::EnableSideEffectsIf { .. }
+                | Instruction::IncrementRc { .. }
+                | Instruction::MakeArray { .. } => {}
+            }
+        }
+
+        trace!("resulting Brillig involved values: {:?}", self.tainted);
+    }
+
+    /// Every Brillig call not properly constrained should remain in the tainted set
+    /// at this point. For each, emit a corresponding warning.
+    fn collect_warnings(&mut self, function: &Function) -> Vec<SsaReport> {
+        let warnings: Vec<SsaReport> = self
+            .tainted
+            .keys()
+            .map(|brillig_call| {
+                SsaReport::Bug(InternalBug::UncheckedBrilligCall {
+                    call_stack: function.dfg.get_call_stack(*brillig_call),
+                })
+            })
+            .collect();
+
+        trace!("making following reports for function {}: {:?}", function.name(), warnings);
+        warnings
+    }
+
+    /// Update sets of value ids that can be traced back to the Brillig calls being tracked
+    fn update_children(&mut self, parents: &[ValueId], children: &[ValueId]) {
+        for (_, tainted_ids) in self.tainted.iter_mut() {
+            for parent in parents {
+                tainted_ids.update_children(parent, children);
+            }
```

Comment:
This method is called as `self.update_children(arguments, results)`, then iterates over each argument calling the [update_children](https://github.com/noir-lang/noir/blob/55f991dc3b4111b768dc81a2e55be2ec6f481efe/compiler/noirc_evaluator/src/ssa/checks/check_for_underconstrained_values.rs#L139-L152) of the tainted structure. 

NB the `children` will be processed multiple times, once for each argument that is already part of of `BrilligTaintedIds::arguments`. It could be more efficient to do the intersection trick on all parents, and if _any_ parent is found then add _all_ children once.

Reply:
This requires creating a HashSet of `parents` on the go, but yes it's probably worth it. Changed